### ISSUE TITLE
Usage of MSTest Runner in AzDO pipelines

### ISSUE THREAD

**Issue by fforjan** [microsoft/testfx#2500]
From #2489, there was a remark from @MarcoRossignoli to use the MStest Runner.

Following a quick look at https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-runner-intro?tabs=continuous-integration#run-and-debug-tests the information seems light.

In our current pipelines, we are doing something like this :
```yml
  - task: VSTest@2
    displayName: 'Run .net 472 tests'
    inputs:
      testSelector: 'testAssemblies'
      testAssemblyVer2: |
        Source\**\net472\*Tests.dll
      codeCoverageEnabled: true
      testRunTitle: '$(Build.DefinitionName) - net472'

  - task: VSTest@2
    displayName: 'Run .net 6 tests'
    inputs:
      testSelector: 'testAssemblies'
      testAssemblyVer2: |
        Source\**\net6.0*\*Tests.dll        
      codeCoverageEnabled: true
      testRunTitle: '$(Build.DefinitionName) - net460'
```

Which means:
it runs all the tests found in our solution, generate 1 test run - meaning 1 test run result, 1 TRX and 1 Coverage file in the test pipelines - for both our .NET Framework and .Net 6

it is not clear from the documentation
```
- task: CmdLine@2
  displayName: "Run Contoso.MyTests"
  inputs:
    script: '.\Contoso.MyTests\bin\Debug\net8.0\Contoso.MyTests.exe'
```

How can we achieve the same result which is one test run for all each frameworks meaning 1 run result, 1 TRX and 1 Coverage for both platforms .net fwk and .net 6  ?
- we have 50+ test projects , some partially only net472 -hello WPF related tests, some both net472 and net6.0 so wildcard are welcome
- in order to simplify our analysis and reports, having one test run instead of 50 TRX and Coverage file is quite important.

**Comment by MarcoRossignoli** [comment]
Thanks for this.

At the moment we don't have a counterpart of VSTest task for the new platform.

The only solution for now is to run all executables using a simple batch script and merge the results at the end if you want to use the runner model. Anyway current MSTest with the runner supports VSTest task so you can use the runner in your local development and run everything with the current VSTest task for now.

Another model the new platform can support is to have 1 testhost(executables) and all the other projects linked. This model is not yet exposed by MSTest but I think that anyway is not what you want because means build all the tests every time you run the "global" test host.

You could run with `dotnet test`  but again the produced assets won't be merged in 1.

I think that we should leverage the `dotnet test` and give to it all the merge feature that VSTest task is bringing and allow a usage like

`dotnet test   --testsModuleFilter:"Source\**\net472\*Tests.dll" --coverage --report-trx`

And if all the below modules registers correct extensions everything should run and at the end we should have a way to merge all the assets.

**Comment by fforjan** [comment]
> Anyway current MSTest with the runner supports VSTest task so you can use the runner in your local development and run everything with the current VSTest task for now.

Are we saying that I could update the pattern on my vstask in the pipeline to check for *.exe and it would run also in the pipeline while on a developer visual studio it would run as a standalone Test runner ?

it seems then we would have 2 different platform which may have different behaviors but could be a good way to allow to move forward

**Comment by MarcoRossignoli** [comment]
> Are we saying that I could update the pattern on my vstask in the pipeline to check for *.exe

Depends if you've mixed netcore/netfx...basically an executable in .NET is and .exe close a .dll and so your filter would not change at all and you task will run always the .dll (currently .NET test are built as exe)
If you've .NET Framework application the story is different because your test will inside the .exe "for real" and so you can pass the .exe as filter and the test platform will run it like a normal dll

So the problem is that a .NET exe is not like a .NET Framework exe and I need to check if MSTest is capable to recognize that it's a .NET exe but try to load the .dll , @Evangelink do you if MSTest supports it? Otherwise we could think to add this capability to have same .exe filter for both platform.

**Comment by fforjan** [comment]
@MarcoRossignoli  nothing urgent on my side, it will take couple of month between our deploymentitem and our WPF issues to be fixed but I'm thinking other may have the same requirements - or maybe we're the only one trying to uprgade from .net framework to .net 'core' :)

**Comment by Evangelink** [comment]
>  @Evangelink do you if MSTest supports it? Otherwise we could think to add this capability to have same .exe filter for both platform.

Not sure to understand what you are trying to achieve. 
We have different entry points when called from VSTest or the runner/new platform, this might be enough.

**Comment by MarcoRossignoli** [comment]
> Not sure to understand what you are trying to achieve.

Let me rephrase, when we build .NET application the result is an test.exe + test.dll MSTest can accept exe and dll as source extension but if we pass the .exe for .NET we cannot "load assembly" it because it's not a real assembly and the IL is inside the test.dll so I was wondering if we pass .exe for both .NET and .NET Framework to the VSTest task that after will pass the source to MSTest we're correctly loading the test.dll and not the test.exe that I think would not work.

**Comment by MarcoRossignoli** [comment]
> @MarcoRossignoli nothing urgent on my side, it will take couple of month between our deploymentitem and our WPF issues to be fixed but I'm thinking other may have the same requirements - or maybe we're the only one trying to uprgade from .net framework to .net 'core' :)

You're not the only here the difference is the interaction between VSTest filter and the new MSTest runner that allows you to run the test "directly" running the exe or in back compat way passing the ".dll" to VSTest or dotnet test like in the past.

**Comment by Evangelink** [comment]
> In our current pipelines, we are doing something like this :
> 
> ```yaml
>   - task: VSTest@2
>     displayName: 'Run .net 472 tests'
>     inputs:
>       testSelector: 'testAssemblies'
>       testAssemblyVer2: |
>         Source\**\net472\*Tests.dll
>       codeCoverageEnabled: true
>       testRunTitle: '$(Build.DefinitionName) - net472'
> 
>   - task: VSTest@2
>     displayName: 'Run .net 6 tests'
>     inputs:
>       testSelector: 'testAssemblies'
>       testAssemblyVer2: |
>         Source\**\net6.0*\*Tests.dll        
>       codeCoverageEnabled: true
>       testRunTitle: '$(Build.DefinitionName) - net460'
> ```

Hi @fforjan, if I am reading fine, you already have a split task (2 calls one for netfx and netcore). If that's the case, then the problem is only for the netfx part because with the runner you are now only producing an `.exe` and not a `.dll` (for netcore the 2 are produced and the exe is only bootstrapping the dll). So I would suggest to change the filter of the netfx to be `Source\**\net472\*Tests.exe` (notice `.exe` instead of `.dll`). Also, there is a typo for the test run title of the net6.0 part (you have written `- net460` when `- net60`.

@nohwnd Can you confirm exes are correctly considered by VSTest?

**Comment by fforjan** [comment]
So if my understand is correct, in that case, on the buidl machine, we would still use the old VsTest environment for unit test as we are using the pipeline but on a developer workstation it would be using the new testrunner environment ?

**Comment by Evangelink** [comment]
Correct! If you want to run in CI with the new runner, you would need to either use `dotnet test` (see https://learn.microsoft.com/dotnet/core/testing/unit-testing-mstest-runner-integrations#dotnet-test---mstest-runner-mode) or calling the exe(s) directly. What you would need is to write a bash/PowerShell script that collects the exe and run them (be careful as the easiest implementation would call exes one after the other instead of running multiple exes in parallel). That's why we would recommend running with `dotnet test`.

**Comment by Evangelink** [comment]
> @Evangelink would you recommend to move away from the VSTest pipeline task and switch to a dotnet test based pipeline ?
It seems then it would provides a better migration path ?

Yes, we would definitely recommend to switch over to `dotnet test` as it seems to be the best solution forward from all discussions we had so far.

> Should [learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-runner-intro?tabs=continuous-integration#run-and-debug-tests](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-mstest-runner-intro?tabs=continuous-integration#run-and-debug-tests)

I agree, we need to update the doc.

> if you want to run it 'in-place' where the code was built, dotnet test would be the preferred option into the CI ?

With `dotnet test` you can also pass some dlls so you don't have to be building in the same place.

**Comment by Evangelink** [comment]
We are going to start working on this pretty soon. There will be some part of code here and some part in https://github.com/dotnet/sdk where `dotnet test` code is. I have created #2577 to track the progress.

Let me move forward by closing this ticket.

**Comment by bt-Knodel** [comment]
In case anyone's looking for a "quick fix" to utilize the ADO VSTest task, I was able to fix this issue by specifying a "pathtoCustomTestAdapters", and using a .NET Framework 4.8 project as the directory ADO searches.  It appears if it finds a .NET Framework test adapter, the .NET 6 tests run just fine.

**Comment by Evangelink** [comment]
@bt-Knodel this would not allow to run tests with the runner but would fallback in VSTest mode. WHich is already working without any trick.
